src/filesystem.c

```diff
@@ -263,7 +263,11 @@ rcutils_calculate_directory_size(const char * directory_path, rcutils_allocator_
       // Skip over local folder handle (`.`) and parent folder (`..`)
       if (strcmp(data.cFileName, ".") != 0 && strcmp(data.cFileName, "..") != 0) {
         char * file_path = rcutils_join_path(directory_path, data.cFileName, allocator);
-        dir_size += rcutils_get_file_size(file_path);
+        if (rcutils_is_directory(file_path)) {
+          dir_size += rcutils_calculate_directory_size(file_path, allocator);
```

Comment:
I'm somewhat concerned about this being a recursive operation.  If you have a very deeply nested directory tree, this could use up quite a lot of stack space (or even hit the default limit).

One way to get around this would be to make this iterative rather than recursive.  We would still need to allocate memory to store which items to visit next, but then it would be from the heap.

What do people think about my concern?

Reply:
I do agree with you. recursive operation could be concern. We might want to get directory recursive iterators at 1st and call `rcutils_calculate_directory_size` for each. this can avoid calling function recursively.

Reply:
> I'm somewhat concerned about this being a recursive operation. If you have a very deeply nested directory tree, this could use up quite a lot of stack space (or even hit the default limit).
>
> One way to get around this would be to make this iterative rather than recursive. We would still need to allocate memory to store which items to visit next, but then it would be from the heap.

Completely agree.  

Maybe we can provide another API function to provide recursive operation support. 
rcutils_calculate_directory_size() API isn't changed.   
Add new API rcutils_calculate_directory_size_with_recursion().  
Real implementation are moved to rcutils_calculate_directory_size_impl().
```c
rcutils_calculate_directory_size(...) {
  rcutils_calculate_directory_size_impl(..., is_recursive = 0);
}

rcutils_calculate_directory_size_with_recursion (...) {
  rcutils_calculate_directory_size_impl(..., is_recursive = 1);
}
```

What do you think ? 

About whether rcutils_calculate_directory_size_with_recursion() is used to replace rcutils_calculate_directory_size() in rosbag2, I will discuss with @Karsten1987.

Reply:
> Recursive operation could be concern. We might want to get directory recursive iterators at 1st and call rcutils_calculate_directory_size for each. this can avoid calling function recursively.

Yes. But it hard to make sure whether the depth of the directory tree need to be added in the future (For other database).   
Maybe we can consider there is a reasonable maximum depth limitation in API function. Such as 30.

Reply:
if i wasn't mistaken, that is not exactly pointed here. I didn't mean the concern is about depth of the tree (size of memory), but location of memory. (not stack but heap) calling the caller function in the callee, this could hit the limit of stack size. So i was thinking that instead of calling the function recursively, get the directory iterators at 1st, then get all file size in `a single` function.

Reply:
> if i wasn't mistaken, that is not exactly pointed here. I didn't mean the concern is about depth of the tree (size of memory), but location of memory. (not stack but heap) calling the caller function in the callee, this could hit the limit of stack size. So i was thinking that instead of calling the function recursively, get the directory iterators at 1st, then get all file size in `a single` function.

Yes, this is exactly the kind of thing I was thinking about.  If we do this iteratively, instead of recursively, then we can both limit the total amount of memory needed and make sure that memory is allocated from the heap rather than from the stack (no matter how deep the structure is).  Something like:

```
std::vector<std::string> directory_names{"/foo"};
int size = 0;
do {
   diriter = get_directory_iter(directory_names[0]);
   for (filename in diriter) {
      if (is_file(filename)) {
        size += filesize(filename);
      } else if (is_dir(filename)) {
        directory_names.push_back(filename);
      }
   }
   directory_names.pop();
} while (directory_names.size() > 0);
```

Separately, we might *also* considering allowing the API user to limit the recursion depth (mostly to limit the time it takes for this operation to complete).

Reply:
@clalancette @fujitatomoya 

Thanks for further explanation.  
I understand.  
I will update codes.